syntax = "proto3";

import "google/protobuf/struct.proto";

message Children {
  repeated NodeResponse nodes = 1;
}

message NodeResponse {
  string id_ = 1;
  oneof sub_node {
    Children children = 2;
    google.protobuf.ListValue list = 3;
  }
}

message PathRequest {
  repeated string path = 1;
}

message SuccessResponse {
  bool success = 1;
}

message CreateRowsRequest {
  repeated string table_path = 1;
  repeated google.protobuf.Struct rows = 2;
}

message CreateColumnsRequest {
  repeated string table_path = 1;
  map<string, google.protobuf.Struct> columns = 2;
}

message UpdateRowRequest {
  repeated string table_path = 1;
  int64 row_id = 2;
  google.protobuf.Struct sub_row = 3;
}

message UpdateColumnRequest {
  repeated string table_path = 1;
  string column_id = 2;
  google.protobuf.Struct column = 3;
}

message IntersectTablesRequest {
  string by_column_id = 1;
  repeated string table1_path = 2;
  repeated string table2_path = 3;
  repeated string new_table_path = 4;
}

message ColumnsResponse {
  google.protobuf.Struct columns = 1;
}

message ColumnResponse {
  google.protobuf.ListValue values = 1;
}

message ValueResponse {
  google.protobuf.Value value = 1;
}

service Tree {
  // Create
  rpc CreateBase(PathRequest) returns (SuccessResponse) {};
  rpc CreateTable(PathRequest) returns (SuccessResponse) {};
  rpc CreateRows(CreateRowsRequest) returns (SuccessResponse) {};
  rpc CreateColumns(CreateColumnsRequest) returns (SuccessResponse) {};

  // Read
  rpc ReadTree(PathRequest) returns (NodeResponse) {};
  rpc ReadBase(PathRequest) returns (NodeResponse) {};
  rpc ReadTable(PathRequest) returns (NodeResponse) {};
  rpc ReadRows(PathRequest) returns (NodeResponse) {};
  rpc ReadRow(PathRequest) returns (NodeResponse) {};
  rpc ReadColumns(PathRequest) returns (ColumnsResponse) {};
  rpc ReadColumn(PathRequest) returns (ColumnResponse) {};
  rpc ReadValue(PathRequest) returns (ValueResponse) {};

  // Update
  rpc UpdateRow(UpdateRowRequest) returns (SuccessResponse) {};
  rpc UpdateColumn(UpdateColumnRequest) returns (SuccessResponse) {};

  // Delete
  rpc DeleteBase(PathRequest) returns (SuccessResponse) {};
  rpc DeleteTable(PathRequest) returns (SuccessResponse) {};
  rpc DeleteRow(PathRequest) returns (SuccessResponse) {};
  rpc DeleteColumn(PathRequest) returns (SuccessResponse) {};

  rpc IntersectTables(IntersectTablesRequest) returns (SuccessResponse) {};
}